File: ios/Kurir/modules/chat/ChatModule.swift
//
//  ChatModule.swift
//  Kurir
//
//  Created by Loren on 19/9/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Mobilex

class ChatModule: Module {
    override func inject() {
        // Worker
        factory(ChatWrapper.self) { r in
            ChatWrapper(userData: r.resolve(UserData.self),
                        repository: r.resolve(MessagesRepository.self)!)
        }
        
        // Actor
        factory(ChatActor.self) { (r, scene: ChatScene) -> ChatActor in
            ChatActor(scene: scene, wrapper: r.resolve(ChatWrapper.self)!)
        }
    }
}

File: android/src/main/java/au/com/crazybean/mobilex/kurir/modules/chat/ChatModule.kt
package au.com.crazybean.mobilex.kurir.modules.chat

import au.com.crazybean.mobilex.kurir.extension.qualifier
import au.com.crazybean.mobilex.kurir.modules.base.ViewModel
import org.koin.androidx.viewmodel.dsl.viewModel
import org.koin.dsl.module

val chatModule = module {
    val qualifier = qualifier<ChatWrapper>()

    // Wrapper via ViewModel
    viewModel(qualifier) {
        ViewModel(ChatWrapper(get(), get()))
    }

    // Actor
    factory { (scene: ChatScene?) ->
        val viewModel = get<ViewModel<ChatWrapper>>(qualifier)
        ChatActor(scene, viewModel.wrapper)
    }
}